File: algorithms/js/860.柠檬水找零.js
/*
 * @lc app=leetcode.cn id=860 lang=javascript
 *
 * [860] 柠檬水找零
 */

// @lc code=start
/**
 * @param {number[]} bills
 * @return {boolean}
 */
var lemonadeChange = function(bills) {
  let five = 0,
    ten = 0;
  for (let i = 0; i < bills.length; i++) {
    if (bills[i] === 5) {
      five++;
    } else if (bills[i] === 10) {
      ten++;
      five--;
    } else {
      if (ten > 0) {
        ten--;
        five--;
      } else {
        five -= 3;
      }
    }
    if (five < 0) {
      return false;
    }
  }
  return true;
};
// @lc code=end

File: algorithms/golang/860.柠檬水找零.go
/*
 * @lc app=leetcode.cn id=860 lang=golang
 *
 * [860] 柠檬水找零
 */

// @lc code=start
func lemonadeChange(bills []int) bool {
	fiveCoins, tenCoins := 0, 0
	for _, bill := range bills {
		switch bill {
		case 5:
			fiveCoins++
		case 10:
			if fiveCoins == 0 {
				return false
			} else {
				fiveCoins--
				tenCoins++
			}
		case 20:
			if fiveCoins > 0 && tenCoins > 0 {
				fiveCoins--
				tenCoins--
			} else if fiveCoins > 2 {
				fiveCoins -= 3
			} else {
				return false
			}
		}
	}
	return true
}

// @lc code=end
